### ISSUE TITLE
Missing `Module#method_added` method

### ISSUE THREAD

**Issue by AlexWayfer** [rubyapi/rubyapi#579]
Hello.

I've noticed that `Module#method_added` method is missing on RubyAPI.org, but exists on Ruby-Doc.org

https://rubyapi.org/2.7/o/module

![image](https://user-images.githubusercontent.com/6510020/95797375-104f7e80-0cf8-11eb-87aa-3af48f356265.png)

https://ruby-doc.org/core-2.7.2/Module.html#method-i-method_added

![image](https://user-images.githubusercontent.com/6510020/95797387-1cd3d700-0cf8-11eb-9dd3-eb4086586778.png)

I think this issue is not related to #346 because source looks like a C code, not Ruby.

**Comment by natematykiewicz** [comment]
```ruby
doc.method_list.each do |method_doc|
  method = {
    name: method_doc.name,
    description: clean_description(doc.full_name, method_doc.description),
    object_constant: doc.full_name,
    method_type: "#{method_doc.type}_method",
    source_location: "#{@release.version}:#{method_path(method_doc)}:#{method_doc.line}",
    alias: {
      path: clean_path(method_doc.is_alias_for&.path, constant: doc.full_name),
      name: method_doc.is_alias_for&.name
    },
    call_sequence: call_sequence_for_method_doc(method_doc),
    source_body: format_method_source_body(method_doc),
    metadata: {
      depth: constant_depth(doc.full_name)
    }
  }

  byebug if doc.full_name == 'Module' # << Added this line

  next if methods.any? { |m| m[:name] == method[:name] && m[:method_type] == method[:method_type] }

  methods << method
end
```

```sh
 ./bin/rake import:ruby\[2.7.2\]
```

```ruby
doc.method_list.map(&:name).sort

["<", "<=", "<=>", "==", "===", ">", ">=", "alias_method", "ancestors", "attr", "attr_accessor", "attr_reader", "attr_writer", "autoload", "autoload?", "class_eval", "class_exec", "class_variable_defined?", "class_variable_get", "class_variable_set", "class_variables", "const_defined?", "const_get", "const_missing", "const_set", "const_source_location", "constants", "constants", "define_method", "deprecate_constant", "freeze", "include", "include?", "included_modules", "inspect", "instance_method", "instance_methods", "method_defined?", "module_eval", "module_exec", "name", "nesting", "new", "prepend", "private_class_method", "private_constant", "private_instance_methods", "private_method_defined?", "protected_instance_methods", "protected_method_defined?", "public_class_method", "public_constant", "public_instance_method", "public_instance_methods", "public_method_defined?", "remove_class_variable", "remove_method", "singleton_class?", "to_s", "undef_method", "used_modules"]
```

I don't see `method_added` in that list.

In #346, Neurogami mentions:
>It's complicated. Getting proper Rdoc for the std-lib on ruby-doc.org has a been a every-evolving challenge for me. So over the years I've crafted assorted scripts to manage correct doc'ing of libs. You have to look a carefully at what files constitute a library, where do they live, and how do you tell rdoc to include them.

Maybe we should look at these lines:
https://github.com/rubyapi/rubyapi/blob/master/lib/ruby_documentation_importer.rb#L25-L38

**Comment by natematykiewicz** [comment]
Searching the source code, I do see that `method_added` is defined on `Kernel` (not `Module`) as `rb_obj_dummy1`.
https://github.com/ruby/ruby/blob/5e120a238904dd1d9e104b5b4cb97cae052c25c8/object.c#L4522

```c
rb_define_private_method(rb_cModule, "method_added", rb_obj_dummy1, 1);
```

Tons of methods are defined like that:

```c
    rb_mKernel = rb_define_module("Kernel");
    rb_include_module(rb_cObject, rb_mKernel);
    rb_define_private_method(rb_cClass, "inherited", rb_obj_dummy1, 1);
    rb_define_private_method(rb_cModule, "included", rb_obj_dummy1, 1);
    rb_define_private_method(rb_cModule, "extended", rb_obj_dummy1, 1);
    rb_define_private_method(rb_cModule, "prepended", rb_obj_dummy1, 1);
    rb_define_private_method(rb_cModule, "method_added", rb_obj_dummy1, 1);
    rb_define_private_method(rb_cModule, "method_removed", rb_obj_dummy1, 1);
    rb_define_private_method(rb_cModule, "method_undefined", rb_obj_dummy1, 1);
```

`rb_obj_dummy1` doesn't do anything other than return nil (it's a callback, so that makes sense).

```c
static VALUE
rb_obj_dummy()
{
    return Qnil;
}

static VALUE
rb_obj_dummy0(VALUE _)
{
    return rb_obj_dummy();
}

static VALUE
rb_obj_dummy1(VALUE _x, VALUE _y)
{
    return rb_obj_dummy();
}
```

Its documentation seems to live in that file.
https://github.com/ruby/ruby/blob/5e120a238904dd1d9e104b5b4cb97cae052c25c8/object.c#L947-L967

I do see more referenced to `method_added` in this file: https://github.com/ruby/ruby/blob/0164ac70a14c3cc17989d76959cf791ec22e8695/vm_method.c

I don't know a ton about C (I've been trying to learn), so I don't really know how this comes into play, but this changelog entry does mention vm_method.c. https://github.com/ruby/ruby/blob/c5eb24349a4535948514fe765c3ddb0628d81004/doc/ChangeLog-2.4.0#L7637-L7641

**Comment by tadman** [comment]
Coincidentally I found that [`remove_const`](https://www.rubydoc.info/stdlib/core/2.7.0/Module:remove_const) is also [missing](https://rubyapi.org/2.7/o/s?q=remove_const).

**Comment by colby-swandale** [comment]
😒 I figured this issue was going to show up at some point but didn't realise how RDoc basically doesn't shows this info to users. 

The TL;DR for this problem is that `method_added ` & `remove_const` are private methods, which by default RDoc (and by extension rubyapi.org also) will skip over.

[docs.ruby-lang.org](https://github.com/ruby/docs.ruby-lang.org/blob/master/Rakefile#L44) and ruby-doc.org are explicitly configuring RDoc to include private methods in the published docs, which I guess we will also have to opt-into as well but preferably with better signage to inform users that certain methods are private or protected.

**Comment by natematykiewicz** [comment]
Shoot.

So, similar to how the attributes have read/write/readwrite tracked, we'll also need to track whether methods/attributes are public/protected/private.

I feel like you'll want iconography to denote that. When displaying methods, do you group by their access level or keep it alphabetical?

**Comment by natematykiewicz** [comment]
Is that #346's problem as well?

**Comment by colby-swandale** [comment]
#346 has a different issue that's related to the `.document` files. I'll post the specific details in the issue.

**Comment by AlexWayfer** [comment]
Thank you @natematykiewicz and @colby-swandale for investigation.

**Comment by isabanin** [comment]
I found another method that's missing: `Module#module_function`. It's available in the official RDoc:

https://ruby-doc.org/core-3.0.0/Module.html#method-i-module_function

**Comment by colby-swandale** [comment]
I've enabled enabled the importer to include private methods into the docs in https://github.com/rubyapi/rubyapi/commit/b633730753272ee1930b6a946d27d5056d126c9b. Let me know if there are still methods not showing up.

**Comment by colby-swandale** [comment]
I will worry about indicating public/protected/private methods in the UI at some point in the future.

**Comment by natematykiewicz** [comment]
I made a new Issue for that, so we don't forget.

**Comment by colby-swandale** [comment]
I'm closing this issue as I believe this has been resolved with https://github.com/rubyapi/rubyapi/commit/b633730753272ee1930b6a946d27d5056d126c9b. Feel free to open a new issue if there's any method still missing.

**Comment by AlexWayfer** [comment]
> I'm closing this issue as I believe this has been resolved with [b633730](https://github.com/rubyapi/rubyapi/commit/b633730753272ee1930b6a946d27d5056d126c9b). Feel free to open a new issue if there's any method still missing.

Created #696, not because the reported method is missing, but due to its wrong documentation.
